src/commons/crypto/signing.rs

```diff
@@ -24,260 +26,321 @@ use rpki::{
+        self.general_signer.read().unwrap().sign(key_id, algorithm, data)
     }
 
-    pub fn sign_rta(&self, rta_builder: &mut rta::RtaBuilder, ee: Cert) -> CryptoResult<()> {
-        let key = ee.subject_key_identifier();
-        rta_builder.push_cert(ee);
+    fn sign_one_off<D: AsRef<[u8]> + ?Sized>(
+        &self,
+        algorithm: SignatureAlgorithm,
+        data: &D,
+    ) -> Result<(Signature, PublicKey), Self::Error> {
+        self.one_off_signer.read().unwrap().sign_one_off(algorithm, data)
+    }
 
-        match self {
-            SignerProvider::OpenSsl(signer) => rta_builder.sign(signer.deref(), &key, None, None),
-            #[cfg(feature = "hsm")]
-            SignerProvider::Dummy(signer) => rta_builder.sign(signer.deref(), &key, None, None),
+    fn rand(&self, target: &mut [u8]) -> Result<(), Self::Error> {
+        let signer = self.general_signer.read().unwrap();
+        if signer.supports_random() {
+            signer.rand(target)
+        } else {
+            self.rand_fallback_signer.read().unwrap().rand(target)
         }
-        .map_err(crypto::Error::signing)
     }
 }
 
+//------------ KrillSigner ---------------------------------------------------
+
+/// High level signing interface between Krill and the Signer backends.
+///
+/// KrillSigner:
+///   - Is configured via the Krill configuration file.
+///   - Maps Result<SignerError> to KrillResult.
+///   - Directs signers to use the RPKI standard key format (RSA).
+///   - Directs signers to use the RPKI standard signature algorithm (RSA PKCS #1 v1.5 with SHA-256).
+///   - Offers a higher level interface than the Signer trait.
 #[derive(Clone, Debug)]
 pub struct KrillSigner {
-    // KrillSigner chooses which signer to use when. The noise of handling the enum based dispatch is handled by the
-    // SignerProvider type defined above, patterned after the existing AuthProvider enum based approach.
-    //
-    // Use Arc references so that we can use refer to the same signer instance more than once if that signer should be
-    // used for multiple purposes, e.g. as both general_signer and one_off_signer in this case.
-    //
-    // Use an RwLock because the Signer trait from the rpki-rs crate uses &mut self for create_key() and destroy_key()
-    // operations. In future we might move the responsibility for locking into the signer so that it can lock only what
-    // actually needs to be locked raet
-
-    // The general signer is used for all signing operations except one off signing.
-    general_signer: Arc<RwLock<SignerProvider>>,
-
-    // As the security of a HSM isn't needed for one off keys, and HSMs are slow, by default this should be an instance
-    // of OpenSslSigner. However, if users think the perceived extra security is warranted let them use a different
-    // Signer for one off keys if that's what they want.
-    one_off_signer: Arc<RwLock<SignerProvider>>,
+    dispatcher: Arc<RwLock<SignerRouter>>,
```

Comment:
A couple of issues with this:
- Firstly, while nameshedding I seem to have become conflicted and used both dispatcher and router. I'm not overly fond of either. The static dispatching could be said to happen in the SignerProvider where it matches on the enum value to "dispatch" to the correct code, not in the SignerRouter. Also the SignerRouter, even if it is considered to do dispatching, doesn't just dispatch, it makes choices about where to dispatch to hence the name "router: rather than dispatcher. However having the name "router" in a routing product for some unrelated purpose makes me uncomfortable as well...

- The RwLock on the whole SignerRouter means that creating a key in a KMIP backend will prevent any action on any other signer at the same time even though they cannot conflict with each other. Ideally there wouldn't even be a RwLock on the individual signers but the use of &mut in the `Signer` trait means we need a lock or else the `mut` leaks out to the owning struct requiring the whole "struct" to be locked at once.

Reply:
I pushed a commit to use consistent naming and to remove the unecessary `Arc<RwLock<..>>` around the `SignerRouter`.